{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VXe_jReb5LZO"
      },
      "source": [
        "# Module 1: Sequence alignment"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IiXL6YGw5LYo"
      },
      "source": [
        "## A. Alignment\n",
        "\n",
        "### Exercise A.1 (Alignment Score)\n",
        "Given the following alignment between two sequences S and T, calculate the alignment score between S\n",
        "and T using each of the given scoring systems:\n",
        "\n",
        "</br>\n",
        "</br>\n",
        "\n",
        "\n",
        "| S  = | A | T | T | C | G | - | T | A | C | T | T | A | G | T | C | C |\n",
        "|------|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
        "| T  = | C | T | T | G | G | C | T | A | A | T | A | G | - | T | C | C |\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "|           | Match | Substitution                              | Indel |\n",
        "|-----------|-------|-------------------------------------------|-------|\n",
        "| Scoring A | 0     | 1                                         | 2     |\n",
        "| Scoring B | 0     | d(A,C) = d(G,C) = 1 d(A,G) = d(T,A) = 1.5 | 2     |\n",
        "| Scoring C | 2     | -1                                        | -1.5  |\n",
        "\n",
        "</br>\n",
        "</br>\n",
        "\n",
        "\n",
        "Select the right answer:\n",
        "\n",
        "**A**: Scoring A = 8 \t\tScoring B = 10 \t\tScoring C = -5\n",
        "\n",
        "**B**: Scoring A = 9 \t\tScoring B = 10 \t\tScoring C = 10\n",
        "\n",
        "**C**: Scoring A = 3 \t\tScoring B = 7.5 \t\tScoring C = 2\n",
        "\n",
        "**D**: Scoring A = 11 \t\tScoring B = 11 \t\tScoring C = 11\n",
        "\n",
        "</br>\n",
        "</br>\n",
        "\n",
        "### Exercise A.2 (Global Alignment: Needleman-Wunsch)\n",
        "Given two sequences: S = ATTCGTACTT and T = CTTAGCTAAT, the following alignment matrix is constructed using the Needleman-Wunsch algorithm with the following scores: mismatch score = -1, gap penalty = -1, and match score = 2.\n",
        "\n",
        "![fig1](img/alignment_matrix.png)\n",
        "\n",
        "\n",
        "\n",
        "</br>\n",
        "\n",
        "**A.2a**\n",
        "\n",
        "Complete the alignment matrix and select the right answer:\n",
        "\n",
        "|    |   |   |   |   |   |    |   |   |   |   |   |\n",
        "|----|---|---|---|---|---|----|---|---|---|---|---|\n",
        "| **A.** |   |   |   |   |   | **B.** |   |   |   |   |   |\n",
        "|    | 4 | 7 | 5 | 5 | 5 |    | 4 | 7 | 5 | 5 | 5 |\n",
        "|    | 3 | 7 | 5 | 4 | 4 |    | 3 | 7 | 5 | 4 | 4 |\n",
        "|    | 1 | 4 | 4 | 7 | 6 |    | 1 | 4 | 4 | 5 | 7 |\n",
        "|**C.** |   |   |   |   |   | **D.** |   |   |   |   |   |\n",
        "|    | 4 | 7 | 6 | 5 | 5 |    | 3 | 6 | 5 | 5 | 5 |\n",
        "|    | 3 | 6 | 6 | 5 | 4 |    | 2 | 5 | 5 | 4 | 4 |\n",
        "|    | 2 | 5 | 5 | 8 | 7 |    | 1 | 4 | 4 | 7 | 8 |\n",
        "\n",
        "</br>\n",
        "\n",
        "**A.2b**\n",
        "\n",
        "Trace back the alignment in the matrix and select the right score:\n",
        "<pre>\n",
        "A: 5  \tB: 6\t  \tC: 7\t  \tD: 8\n",
        "</pre>\n",
        "\n",
        "**A.2c**\n",
        "\n",
        "Write down the alignment result (similar to the example given in Exercise A.1).\n",
        "\n",
        "**A.2d**\n",
        "\n",
        "There are many implementations of the Needleman-Wunsch algorithm available. Here, we use the implementation in the minineedle python package. Install the minineedle package using `!pip install minineedle`. Then, use the function NeedlemanWunsch to reproduce your results in the previous parts of this exercise. You will need to provide the match and mismatch scores and set the gap penalty.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GmYn8RwAC02p",
        "outputId": "a4bacd59-6002-4e84-96de-e035aad2e275"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Collecting minineedle\n",
            "  Downloading minineedle-3.1.5-py3-none-any.whl.metadata (5.5 kB)\n",
            "Downloading minineedle-3.1.5-py3-none-any.whl (19 kB)\n",
            "Installing collected packages: minineedle\n",
            "Successfully installed minineedle-3.1.5\n"
          ]
        }
      ],
      "source": [
        "!pip install minineedle\n",
        "from minineedle import needle, core"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Bpn0vStsgOcn",
        "outputId": "3d667bbe-d58f-45a1-8304-82a375e6ee00"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Alignment of SEQUENCE 1 and SEQUENCE 2:\n",
            "\t-------ATTCGTACTT\n",
            "\tCTTAGCTA-----A--T\n",
            "\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "20"
            ]
          },
          "execution_count": 2,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "S = 'ATTCGTACTT' # source sequence\n",
        "T = 'CTTAGCTAAT' # target sequence\n",
        "alignment = needle.NeedlemanWunsch(S, T)\n",
        "gap_penalty = -1\n",
        "mismatch_score = -1\n",
        "match_score = 2\n",
        "alignment.change_matrix(core.ScoreMatrix(match=match_score, miss=mismatch_score, gap=gap_penalty))\n",
        "alignment.align()\n",
        "print(alignment)\n",
        "alignment.get_score()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wvPJ55lTcYq_"
      },
      "source": [
        "**A.2e**\n",
        "\n",
        "In the previous example we used a gap_penalty of -1. What happens if we set the gap penalty to 0? What does the alignment look like?\n",
        "\n",
        "###Exercise A.3 (Global Alignment: Needleman-Wunsch)\n",
        "\n",
        "UniProt Knowledgebase (UniProtKB) is a database of functional information on proteins, which allows you to lookup the amino acid sequence for a given protein. Find the amino acid sequences for the protein sequences below:\n",
        "\n",
        "| Protein                  | Accession Number |\n",
        "|--------------------------|------------------|\n",
        "| Human hemoglobin alpha   | P69905           |\n",
        "| Chimp hemoglobin alpha   | P69907           |\n",
        "| Chicken hemoglobin alpha | P01994           |\n",
        "| Shark hemoglobin alpha   | P02021           |\n",
        "| Human hemoglobin beta    | P68871           |\n",
        "| Human myoglobin          | P02144           |\n",
        "\n",
        "Now use the `NeedlemanWunsch` function from minineedle to align each of the following proteins to the human hemoglobin alpha: chimp hemoglobin alpha, chicken hemoglobin alpha, shark hemoglobin alpha, human hemoglobin beta, and human myoglobin. Use the default scoring matrix.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LQtkRFdFgOdf",
        "outputId": "9b5e13a5-95bb-42bd-a5c3-848565015698"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "3"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "S = '...' # paste your protein sequence 1 here\n",
        "T = '...' # paste your protein sequence 2 here\n",
        "alignment = needle.NeedlemanWunsch(S, T)\n",
        "alignment.align()\n",
        "alignment.get_score()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Pti7A7MEdFNJ"
      },
      "source": [
        "Which is the most similar protein to human hemoglobin alpha?\n",
        "\n",
        "**A**: Chimp hemoglobin alpha\n",
        "\n",
        "**B**: Chicken hemoglobin alpha\n",
        "\n",
        "**C**: Shark hemoglobin alpha\n",
        "\n",
        "**D**: Human hemoglobin beta\n",
        "\n",
        "**E**: Human myoglobin"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NtvLkFuYdQ2R"
      },
      "source": [
        "### Exercise A.4 (Local Alignment: Smith Waterman)\n",
        "\n",
        "Construct the alignment matrix V for S and T, where S = ATTCGTACTT and T = ATTA using the Smith-Waterman algorithm with a mismatch score of -1, a match score of 2 and a gap penalty of -1.\n",
        "\n",
        "**A.4a**\n",
        "\n",
        "What is the value of V(4,10)?\n",
        "\n",
        "<pr>\n",
        "A: 7\t\tB: 6\t\tC: 0\t\tD: 4\n",
        "</pr>\n",
        "\n",
        "**A.4b**\n",
        "\n",
        "Trace back the alignment in the matrix and select the right score:\n",
        "\n",
        "<pr>\n",
        "A: 7\t\tB: 6\t\tC: 0\t\tD: 4\n",
        "</pr>\n",
        "\n",
        "**A.4c**\n",
        "\n",
        "Write down the alignment result.\n",
        "\n",
        "**A.4d**\n",
        "\n",
        "Use the minineedle function `SmithWaterman` to reproduce you results. You will need to provide the match and mismatch scores and set the penalties for starting and extending a gap.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9sD5W99hhzeT",
        "outputId": "ae57d72f-aca3-4b30-99b0-d46a6486e717"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Alignment of SEQUENCE 1 and SEQUENCE 2:\n",
            "\tATTCGTACTT\n",
            "\tA-T--TA---\n",
            "\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "14"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "from minineedle import smith, core\n",
        "\n",
        "S = 'ATTCGTACTT' # source sequence\n",
        "T = 'ATTA' # target sequence\n",
        "alignment = smith.SmithWaterman(S, T)\n",
        "alignment.change_matrix(core.ScoreMatrix(match=match_score,\n",
        "miss=mismatch_score, gap=gap_penalty))\n",
        "alignment.align()\n",
        "print(alignment)\n",
        "alignment.get_score()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9Bf1SRiTdkUo"
      },
      "source": [
        "# B.Databases\n",
        "\n",
        "There is a large collection of publicly available biological databases, many of which are useful for bioinformatics analysis and interpretation of results. In Exercise A.3 you already used one of them, UniProtKB, to look up protein sequences. See [list of biological databases](https://en.wikipedia.org/wiki/List_of_biological_databases) for an extensive overview.\n",
        "\n",
        "### Exercise B.1\n",
        "\n",
        "Suppose you have employed sequencing to identify potentially oncogenic mutations in the DNA of a patient with Acute Myeloid Leukemia. For this purpose, you have compared the DNA of the cancerous cells with DNA obtained from a non-tumor sample (e.g. buccal swap) of the same patient, and identified a nucleotide position that is different in the tumorous sample as compared to the control sample. This somatic mutation is situated in a gene encoding the DNMT3A protein. To learn more about this potential mutation, you decide to review information available in online databases.\n",
        "\n",
        "**B.1a Query UCSC**\n",
        "\n",
        "A lot of information is aggregated in so-called Genome Browsers. One of these is the UCSC:\n",
        "\n",
        "Do:\n",
        "\n",
        ">  [1] Surf to [ucsc](https://genome-euro.ucsc.edu/)\n",
        "  \n",
        ">  [2] Click 'Genome browser' under 'Tools'\n",
        "\n",
        ">  [3] Select human genome in 'human assembly GRCh37 / hg19'\n",
        "\n",
        ">  [4] Enter 'DNMT3A' in the 'Position/Search Term'\n",
        "\n",
        ">  [5] You should see a list of possible hits, categorized by gene definitions from UCSC itself (UCSC) or external databases (RefSeq Genes, etc). Click:\n",
        "'DNMT3A (uc002rgd.4) - chr2:25455830-25565459 - Homo sapiens DNA (cytosine-5-)-methyltransferase 3 alpha (DNMT3A), transcript variant 1, mRNA'\n",
        "\n",
        ">  [6] You should now see a snapshot of the genome with various gene isoform definitions of DNMT3A, including the definition you selected highlighted in a black box under\n",
        "'UCSC genes (RefSeq, GenBank, CCDS, Rfam, tRNAs & Comparative Genomics)' at the top\n",
        "\n",
        "Answer:\n",
        "\n",
        ">  [1] What is a genome assembly?\n",
        "  \n",
        ">  [2] On what chromosome arm is DNMT3A encoded?\n",
        "\n",
        ">  [3] How many gene isoforms are listed for DNMT3A under 'UCSC genes (RefSeq, GenBank, CCDS, Rfam, tRNAs & Comparative Genomics)'?\n",
        "\n",
        "\n",
        "**B.1b Gene structure**\n",
        "\n",
        "Isoform-centric information can also be obtained in this browser:\n",
        "\n",
        "Do:\n",
        "\n",
        "> [1] Click on the highlighted (black box) DNMT3A gene isoform\n",
        "\n",
        "Answer:\n",
        "\n",
        "> [1] What is the full name of DNMT3A?\n",
        "\n",
        "> [2] What is its function?\n",
        "\n",
        "> [3] How many exons does this isoform have? How many exons are coding? What are the other?\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
